### DISCUSSION TITLE
How to understand core architect DNN?

### DISCUSSION THREAD

**Discussion by Mostafa-Moafi** [dnnsoftware/Dnn.Platform#5589] [Q&A]
Hi 
I need to understand the core architect DNN.
How can I?

**Comment by valadas** [comment]
All of the source code is here, there is also https://docs.dnncommunity.org/ which has some more "human readable" information.
Basically it is built on a mix of .Net Standard and .Net Framework, supports webforms, MVC and webAPI module pipelines. There is a ton of extension points for various things like theming, modules, authentication, etc.

If you have any specific questions, feel free to ask.

**Reply by Mostafa-Moafi** [reply]
Thanks, Valadas
I need to understand caching settings in DNN. I mean All caching configs such as Caching Provider or Module Cache Provider and all of the things in the performance setting in the persona bar.

**Reply by Mostafa-Moafi** [reply]
And how to active UsersOnline in the portal?

**Reply by valadas** [reply]
For caching, there are basically 3 things:
1. Pages: provider driven and 3 providers are supplied. File, Database and Memory.
2. For module level cache, it can use File or Memory cache. Modules define their default cache time in the installer manifest and in individual module settings, one can override that default.
3. There is also a DataCache class that allows persisting some data in memory for custom extensions like large datasets that don't change often.

> Other performance settings in the PersonaBar

The other thing that is often used is Client Resource Management, basically, if turned on, the static files (.js, .css) requested by extensions will be de-duplicated (say 4 modules request jQuery). Then if you turn on "Enable Composite Files" on top of that the files will be concatenated to make it a single HTTP request of larger size instead of loading multiple small files (there are pros and cons to this). And if that is turned on, then you can optionally also minify CSS and js, which will remove comments, change variable names to shorter names, etc.

For UsersOnline, there was support for this in DNN but it was all marked deprecated in v8 so it will be gone in v10. It was a massive performance issue. You do have in the users table the last login date, etc. that might be of some use. But analytics engines might be a better solution that won't affect DNN performance

**Reply by Mostafa-Moafi** [reply]
Thank you very much.
Please explain to me Cache Setting, Authenticated Cacheability, and Unauthenticated Cacheability.
For Users Online, I need to detect admin or other roles and the login date doesn't work for me because I need the information at the moment. Do you have other suggestions for me?

**Comment by valadas** [comment]
You can have different caching settings for unauthenticated users (visitors) vs authenticated users basically.

As for UserOnline I am afraid this would need a 3rd party solution but I am not aware of any...

**Comment by thienvc** [comment]
> For UsersOnline, there was support for this in DNN but it was all marked deprecated in v8 so it will be gone in v10. It was a massive performance issue. You do have in the users table the last login date, etc. that might be of some use. But analytics engines might be a better solution that won't affect DNN performance

I suggest keeping the UserOnline feature but adding clear warnings that affect performance?
Because I guess DNN is used more in Intranet networks for a form of management application
